---
title: "Load TPMs matrix and compare with expression, and splicing ones"
author: "ToniG"
output: html_notebook
date: "09/27/2021"

---

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)
```
Load TPMs matrix coming from Kallisto
```{r begin}
library(dplyr)
library(tidyr)
library(ggplot2)

df3<-read.delim("/home/proteome/bioinformatics/Data/Storage/Molecular/GeneSplicing/WholeBlood/2017_IMX_P4RNAseq_NovaSeqK_IA/01_Abundance/TPM_filtered_all_data.tpm")
rownames(df3)<-df3$X
df3<-df3[,-1]
head(df3)

##before follow on, just change AGAIN (ISOML), prior to correlate later with expression
pheno<-readRDS("/home/proteome/bioinformatics/Data/Storage/Molecular/GeneExpression/WholeBlood/2017_IMX_P4RNAseq_NovaSeqK_IA/Metadata/Annotation_RNAseq_Table.rds")
####same order as splicing matrix
pheno<-pheno[pheno$celgIdAll %in% colnames(df3),]
pheno<-pheno[match(colnames(df3),pheno$celgIdAll),]
colnames(df3)<-pheno$donaId
head(df3)
```

###PCA
```{r PCA}

pca.res <- FactoMineR::PCA(t(df3), scale.unit = F, graph = F, ncp = 20)###no scale


# Scree plot per veure variança de cada component
factoextra::fviz_eig(pca.res)
```
```{r PCA2}

factoextra::fviz_pca_ind(pca.res, axes = c(1,2), habillage=as.factor(pheno$imid),geom="point")+ coord_fixed()
```
Nearly all variance on PC1
```{r}

factoextra::fviz_pca_ind(pca.res, axes = c(1,2), habillage=as.factor(pheno$Sex),geom="point")##no effect
```
No Sex effect 
```{r}

factoextra::fviz_pca_ind(pca.res, axes = c(1,2), habillage=as.factor(pheno$batch),geom="point")
##not so clearbatch effect
```
Clear batch effect

```{r}

# Age cuts
pheno$age_cut <- as.numeric(cut(pheno$age, breaks=10))

# Create variable for age plotting
factoextra::fviz_pca_ind(pca.res, axes = c(1,2), habillage=as.factor(pheno$age_cut),geom="point")##no effect

```
No clear age

##outliers
```{r}
outs.class<-apply(data.frame(pca.res$ind$coord)[,c(1,2)], 2, function(x) which( (abs(x - median(x)) / mad(x)) > 3 )) %>%
  Reduce(union, .)

outliers<-rownames(data.frame(pca.res$ind$coord)[,c(1,2)])[outs.class]
##1

pheno$out<-ifelse(rownames(pheno) %in% outliers,"YES","NO")
#plot
factoextra::fviz_pca_ind(pca.res, axes = c(1,2), habillage=as.factor(pheno$out),geom="point")
```

Correlations Gene Expression PCs

```{r Corrs}

###Gene expression components

gene.exprs<-readRDS("/home/proteome/bioinformatics/Data/Storage/Molecular/GeneExpression/WholeBlood/2017_IMX_P4RNAseq_NovaSeqK_IA/DerivedFeatures/Quantitative/Principal_Components/Results/Classic.rds")
  
####get components of PSI
transcript.exprs <- data.frame(pca.res$ind$coord)
gene.exprs<-gene.exprs[match(rownames(transcript.exprs),rownames(gene.exprs)),]

# Compute correlation matrix
cor.df <- cor(transcript.exprs[,1:10], gene.exprs[,1:10], use = "pairwise.complete.obs")
corrplot::corrplot(cor.df, method = "number", tl.cex = 0.7, number.cex = 0.7)
```
```{r Corrs_psi}

###Gene expression components

gene.psi<-readRDS("/home/proteome/bioinformatics/Data/Storage/Molecular/GeneSplicing/WholeBlood/2017_IMX_P4RNAseq_NovaSeqK_IA/DerivedFeatures/Quantitative/Principal_Components/Results/Classic.rds")
  
####get components of PSI
gene.psi<-gene.psi[match(rownames(transcript.exprs),rownames(gene.psi)),]

# Compute correlation matrix
cor.df <- cor(transcript.exprs[,1:10], gene.psi[,1:10], use = "pairwise.complete.obs")
corrplot::corrplot(cor.df, method = "number", tl.cex = 0.7, number.cex = 0.7)
```
###Correlation with other variables
###numerical
```{r}

###find which dimension is more associated with sex and age

raw.sampletable<-read.csv("/home/proteome/bioinformatics/Data/Storage/Molecular/GeneExpression/WholeBlood/2017_IMX_P4RNAseq_NovaSeqK_IA/Metadata/Old/Final_Table_GSL.ID.csv")
raw.sampletable<-raw.sampletable[match(rownames(transcript.exprs),raw.sampletable$donaId),]

#(batch, RIN, depth, plate) i vs biològiques (IMID, sexe, edat, activitat). 

cor.df2 <- cor(transcript.exprs[,1:10], raw.sampletable[,c("age","rna.RIN","Fragment.Size","KAPA.nM.","Reads")], use = "pairwise.complete.obs")
corrplot::corrplot(cor.df2, method = "number", tl.cex = 0.7, number.cex = 0.7)
```
```{r}
####put here categorical metric to test with
##https://stats.stackexchange.com/questions/119835/correlation-between-a-nominal-iv-and-a-continuous-dv-variable/124618#124618
####numeric nominal
###Boxplot

library(ggpubr)
library(gridExtra)

metric<-c("Sex","batch","imid")

###build a matrix with n rwos, Dimensions, and 4 columns, metrics

my.cors<-matrix(nrow = 10,ncol = 3)
my.ps<-matrix(nrow = 10,ncol = 3)

for( j in 1:10){ print(j)
  
  for( i in 1:length(metric)){ 
  
  numer<-transcript.exprs[,j]
  nomin<-pheno[,metric[i]]
  
  ##coefficient of determination
  r_squared <- summary(lm(numer ~ nomin))$r.squared
  pval<-broom::glance(summary(lm(numer ~ nomin)))$p.value
  ##multiple correlation coefficent
  assoc<-sqrt(r_squared)
  
  ###put it on matrix
  my.cors[j,i]<-assoc
  my.ps[j,i]<-pval
  
  ####
  
  
  }

}
library(corrplot)

rownames(my.cors)<-c("Dim.1","Dim.2","Dim.3","Dim.4","Dim.5","Dim.6","Dim.7","Dim.8","Dim.9","Dim.10")
colnames(my.cors)<-c("Sex","batch","RESP")


corrplot::corrplot(my.cors, method = "number", tl.cex = 0.7, number.cex = 0.7)



```
##Cell proportions
```{r}

###find which dimension is more associated with sex and age

cells<-readRDS("/home/proteome/bioinformatics/Data/Storage/Molecular/GeneExpression/WholeBlood/2017_IMX_P4RNAseq_NovaSeqK_IA/DerivedFeatures/Quantitative/Cell_Estimates/cell_proportions.rds")

##ISOML again
pheno<-pheno[match(rownames(cells),pheno$celgIdAll),]

rownames(cells)<-pheno$donaId
cells<-cells[match(rownames(transcript.exprs),rownames(cells)),]
main_cells <- c("Granulocytes", "B cells (CD19+)", "CD4+ T cells", "CD8+ T cells", "Monocytes (CD14+)", "NK cells (CD3- CD56+)")


cor.df2 <- cor(transcript.exprs[,1:10], cells[,colnames(cells) %in% main_cells], use = "pairwise.complete.obs")
corrplot::corrplot(cor.df2, method = "number", tl.cex = 0.7, number.cex = 0.7)
```



Follow analysis with genes
```{r}
###get DEG and DTE and check if they are enriched
library(fgsea)

deg<-readRDS("/home/proteome/bioinformatics/Projects/Internal_Projects/2021_IMX_IAP4_Mining/Results/GeneExpression/WholeBlood/01_Features/01_clinicalAssociation/01_caseCtrl/SLE_SABG.rds")
deg<-deg[deg$FDR<0.05,]
rank.deg<-deg$logFC
names(rank.deg)<-rownames(deg)
rank.deg<-rank.deg[order(rank.deg,decreasing = T)]
####get dte to test

dte<-readRDS("/home/proteome/bioinformatics/Projects/Internal_Projects/2021_IMX_IAP4_Mining/Results/TranscriptExpression/WholeBlood/01_Features/01_clinicalAssociation/01_caseCtrl/DTE_FULL_SLE.rds")
dte<-dte[dte$qval<0.05,]


genes_dte<-unique(dte$ext_gene)
my.list<-list(genes_dte)




fgseaRes <- fgsea(pathways = my.list, 
                  stats    = rank.deg,
                  eps      = 0.0,
                  minSize  = 15,
                  maxSize  = 800)

plotEnrichment(genes_dte,
               rank.deg) + labs(title="DTE enriched in DEG pval<2.2e-16")

###DE stats

##How many genes are we gaining?
##How man genes are we losing?
##Are we gaining statistical power?

##gain

overlap<-length(rownames(deg[rownames(deg) %in% genes_dte, ]))/dim(deg)[1]
loss<-deg[!(rownames(deg) %in% genes_dte), ]
gain<-dte[!(dte$ext_gene %in% rownames(deg)), ]
gain_genes<-unique(gain$ext_gene)
loss_genes<-unique(rownames(loss))

genesid<-gain_genes
genesid<-loss_genes

genesid<-genesid[ genesid != "" ]##remove empty elements from a vector
genesid <- genesid[!is.na(genesid)]

eg<-bitr(genesid, fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Hs.eg.db")


ego2 <- enrichGO(gene         = eg$ENTREZID,
                 OrgDb         = org.Hs.eg.db,
                 keyType       = 'ENTREZID',
                 ont           = "BP",
                 pAdjustMethod = "BH",
                 pvalueCutoff  = 0.1,
                 qvalueCutoff  = .05,
                 readable=T
)

dotplot(ego2, title="SLE vs CTRL Gain in DTE",showCategory=30)



##One idea will be to get the pathways and do same analysis as we did in QC for HSIA for RiboZERo, se
##how many of them overlap in the rank of original 

```
```{r,echo=TRUE, fig.keep='all'}


###about the statistical power, get transcript per each gene, the one with more adjusted pval, and plot corr
overlap_genes<-rownames(deg[rownames(deg) %in% genes_dte, ])

ov_dte<-dte[dte$ext_gene %in% overlap_genes,]


library(data.table)
data <- data.table(ov_dte)
data<-data[,.SD[which.max(qval)],by=ext_gene]

deg$ext_gene<-rownames(deg)

genes2plot<-merge(deg[,c("ext_gene","FDR")],data[,c("ext_gene","qval")],by="ext_gene")

mod1 <-lm(as.numeric(-log10(genes2plot$FDR))~as.numeric(-log10(genes2plot$qval)))## 
modsum<-summary(mod1)

r2 <- cor.test(as.numeric(-log10(genes2plot$FDR)),as.numeric(-log10(genes2plot$qval)),method="pearson")$estimate
my.p<-cor.test(as.numeric(-log10(genes2plot$FDR)),as.numeric(-log10(genes2plot$qval)),method="pearson")$p.value
my.p<-signif(my.p, digits=3)


mylabel = bquote(italic(R)^2 == .(format(r2, digits = 3)))
plot(as.numeric(-log10(genes2plot$FDR))~as.numeric(-log10(genes2plot$qval)), xlab="DEG ", ylab="DTE", pch=20,col="grey40")
 #    ,xlim=c(0,10000),ylim=c(0,10000))
#  plot(as.numeric(cnts[,1])~as.numeric(cnts[,2]), main=paste("Correlation","Pulse Replicates Global", "p-value", my.p, sep=" "), xlab="Ctrl1", ylab="Ctrl2", pch=20)

abline(mod1, col="red")

legend('topleft', legend = mylabel, bty = 'n')
legend(x=-6,y=40, legend = paste("pval=",my.p, sep=""),bty = 'n')

library(ggpubr)

####

genes2plot$DEG<- -log10(genes2plot$FDR)
genes2plot$DTE<- -log10(genes2plot$qval)

ggscatter(
  genes2plot, x = 'DEG', y = 'DTE', color = 'blue', 
  add = "reg.line", conf.int = FALSE
  )+
  stat_cor(aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")), show.legend = FALSE) 
#+
  #stat_regline_equation(label.x = 3, label.y = 32)


lm_eqn <- function(df){
    m <- lm(DEG ~ DTE, genes2plot);
    eq <- substitute(italic(y) == a + b %.% italic(x)*","~~italic(r)^2~"="~r2, 
         list(a = format(unname(coef(m)[1]), digits = 2),
              b = format(unname(coef(m)[2]), digits = 2),
             r2 = format(summary(m)$r.squared, digits = 3)))
    as.character(as.expression(eq));
}


p <- ggplot(data = genes2plot, aes(x = DEG, y = DTE)) +
            geom_point(alpha=0.8,color="red")+
            geom_smooth(method = "lm", se=FALSE, color="black", formula = y ~ x) 


p1 <- p + geom_text(x = 5, y = 22, label = lm_eqn(my.df2), parse = T,show.legend = FALSE,color="black")

p1

delta<-cliff.delta(genes2plot$DEG,genes2plot$DTE,return.dm=TRUE)

```
```{r,echo=TRUE, fig.keep='all'}

h1 <- hist(genes2plot$DEG, breaks = 20)
h2 <- hist(genes2plot$DTE, breaks = 20)

par(mar=c(5.1, 4.1, 4.1, 8.1), xpd=TRUE)
c1 <- rgb(255,51,51,max = 255, alpha = 80, names = "lt.blue")
c2 <- rgb(51,255,51, max = 255, alpha = 80, names = "lt.pink")

plot(h1, col = c1, xlab = "-log10(FDR)",ylab="Counts",main = NULL)
plot(h2, col = c2, add = T)

legend(30, 37,c("DEG", "DET"), fill = c(c1, c2), title = "Set")
```
Get pathways from DTE and overlap in bins of DGE

```{r}
genesid<-deg$Feature
genesid<-genesid[ genesid != "" ]##remove empty elements from a vector
genesid <- genesid[!is.na(genesid)]

eg<-bitr(genesid, fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Hs.eg.db")


ego2 <- enrichGO(gene         = eg$ENTREZID,
                 OrgDb         = org.Hs.eg.db,
                 keyType       = 'ENTREZID',
                 ont           = "BP",
                 pAdjustMethod = "BH",
                 pvalueCutoff  = 0.1,
                 qvalueCutoff  = .05,
                 readable=T
)

paths.deg<-as.data.frame(ego2)
###rank them
rank.path.deg<-paths.deg$qvalue
names(rank.path.deg)<-paths.deg$ID


overlap.paths<-enrich.21$ID[enrich.21$ID %in% enrich.18$ID]
my.list<-list(overlap.paths)
```

